# Aufgabe 6 - CNN - Backpropagation: Convolution

Dieses Notebook thematisiert die Backpropagation durch eine Convolution in Convolutional Neural Networks (CNNs).

Ziel ist es, die Funktionsweise der Convolution während der Backpropagation genau zu analysieren und nachzuimplementieren.

<font color="#aa0000">**Hinweis:**</font>
Dieses Notebook enthält Praktikumsaufgaben ([P6.1](#praktikum), [P6.2](#praktikum2), [P6.3](#praktikum3)). Erweitern Sie das Notebook geeignet und speichern Sie das ausgeführte Notebook erneut ab (File &rarr; Download as &rarr; Notebook). Reichen Sie abschließend das heruntergeladene Notebook im zugehörigen [Moodle-Kurs](https://moodle2.tu-ilmenau.de/course/view.php?id=4366) ein.

**Die Einreichungsfrist finden Sie im Moodle-Kurs.**

### Inhaltsverzeichnis
- [(f) Implementierung der Convolution mit NumPy](#f)
    - [Berechnung der Größe der Output Feature Maps](#output_shape)
    - [Berechnung der Convolution](#funktion_conv)
    - [Berechnung der Gradienten](#gradienten)
- [Praktikumsaufgabe P6.1 zur Strided Convolution](#praktikum)
- [Praktikumsaufgabe P6.2 zur Convolution mit Dilation & Padding](#praktikum2)
- [Praktikumsaufgabe P6.3 zur Convolution mit 1x1 Filtern & Gruppen ](#praktikum3)

<hr style="border-width: 5px">

### Vorbereitung
Wichtige Ergebnisse können während der Bearbeitung überprüft werden. Grundvoraussetzung hierfür ist, dass Sie das Paket `tui-dl4cv` <font color="#aa0000">installieren bzw. aktualisieren</font> und anschließend importieren.

Für die Installation stehen Ihnen zwei mögliche Wege zur Verfügung.

**(1) Installation direkt in diesem Notebook:**
Führen Sie den nachfolgenden Code-Block aus.

In [1]:
import sys

print(f"Automatically install package for '{sys.executable}'")
!{sys.executable} -m pip install tui-dl4cv \
    --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" \
    --no-cache --upgrade

Automatically install package for '/usr/bin/python3'
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://2022ws:****@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple
DEPRECATION: The HTML index page being used (https://nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple/tui-dl4cv/) is not a proper HTML 5 document. This is in violation of PEP 503 which requires these pages to be well-formed HTML 5 documents. Please reach out to the owners of this index page, and ask them to update this index page to a valid HTML 5 document. pip 22.2 will enforce this behaviour change. Discussion can be found at https://github.com/pypa/pip/issues/10825


ODER

**(2) Manuelle Installation über die Konsole:**
Öffnen Sie eine Konsole ("Anaconda Prompt" unter Windows) und führen Sie folgenden Befehl aus:
```text
pip install tui-dl4cv --extra-index-url "https://2022ws:xXCgQHZxxeNYchgryN7e@nikrgl.informatik.tu-ilmenau.de/api/v4/projects/1730/packages/pypi/simple" --no-cache --upgrade
```

**Führen Sie abschließend folgenden Code-Block aus, um das Paket verwenden zu können.**
Während der Bearbeitung können Sie nun Ihre Ergebnisse mithilfe der Funktion `interactive_check` überprüfen. Die Funktionsaufrufe sind bereits an den entsprechenden Stellen im Notebook enthalten.

In [2]:
import tui_dl4cv.cnn

# noetige Erweiterung, damit Variablen aus diesem Notebook automatisch ueberprueft werden koennen
def interactive_check(name, **kwargs):
    tui_dl4cv.cnn.interactive_check(name, globals(), **kwargs)

from tui_dl4cv.cnn import print_tensors

<hr style="border-width: 5px">

<a name="f"></a>
### (f) Implementieren Sie die Convolution mithilfe von NumPy und reproduzieren Sie die Ergebnisse der Forward Propagation und der Backpropagation für die letzten beiden Convolutions aus Teilaufgabe (a).

---
Pakete importieren:

In [3]:
# NumPy
import numpy as np

---
<a name="output_shape"></a>
**Zunächst muss eine Funktion implementiert werden, mit deren Hilfe die Größe der Output Feature Maps in Abhängigkeit der verwendeten Parameter der Convolution berechnet werden kann.**
Das Wissen über die Größe der Output Feature Maps ermöglicht es, später ausreichend Speicher für das Ergebnis der Convolution zu allokieren. Zudem gibt sie Aufschluss darüber, wie viele Aufpunkte in der Convolution berechnet werden müssen.
<br>
<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Die Formel zur Berechnung der Größe der Output Feature Maps in Abhängigkeit aller behandelten Parameter der Convolution wird in der Vorlesung und der Übung thematisiert.
</div>

*Implementierung:*

In [4]:
def get_output_shape(input_shape, kernel_size, stride=(1, 1), padding=(0, 0), dilation=(1, 1)):
    assert all(isinstance(p, tuple) for p in locals().values()), "Alle Parameter als Tuple erwartet"

    # Hoehe `h_out` bestimmen (erstes Element in den Tuplen)
    h_out = (input_shape[0] + 2* padding[0] - dilation[0]*(kernel_size[0]-1)-1) / stride[0] +1 # bitte Code ergaenzen <---------------- [Luecke (1)]

    # Breite `w_out` bestimmen (zweites Element in den Tuplen)
    w_out = (input_shape[1] + 2* padding[1] - dilation[1]*(kernel_size[1]-1)-1) / stride[1] +1 # bitte Code ergaenzen <---------------- [Luecke (2)]

    # Ergebnisse durch Cast zu Int abrunden und zurueckgeben
    return int(h_out), int(w_out)

*Überprüfung:*

Zur Überprüfung können die Größen der Output Feature Maps in der Aufgabenstellung berechnet werden.

In [5]:
# Convolution 1
print("Output Shape Conv1:\n",
      get_output_shape(input_shape=(8, 8), kernel_size=(2, 2), stride=(2,2)))

# Convolution 2
print("Output Shape Conv2:\n",
      get_output_shape(input_shape=(4, 4), kernel_size=(2, 2)))

# Convolution 3
print("Output Shape Conv3:\n",
      get_output_shape(input_shape=(3, 3), kernel_size=(3, 3)))

# zusaetzlich komplexer Fall mit Stride, Padding und Dilation:
print("Output Shape:\n",
      get_output_shape(input_shape=(10, 11),
                       kernel_size=(3, 4),
                       stride=(2, 1),
                       padding=(0, 2),
                       dilation=(1, 2)))

Output Shape Conv1:
 (4, 4)
Output Shape Conv2:
 (3, 3)
Output Shape Conv3:
 (1, 1)
Output Shape:
 (4, 9)


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
Output Shape Conv1:
 (4, 4)
Output Shape Conv2:
 (3, 3)
Output Shape Conv3:
 (1, 1)
Output Shape:
 (4, 9)
</code>
</details>

---
<a name="funktion_conv"></a>
**Anschließend kann eine Funktion zur Berechnung der Convolution implementiert werden.**

Beachten Sie:
- Es handelt sich bei den betrachteten beiden Convolutions um Standard Convolutions, daher sollen zunächst keine zusätzlichen Parameter einbezogen werden.
- Der Input Tensor ist 4-dimensional und enthält eine zusätzliche Achse für die Beispiele im Batch.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende Funktionen könnten für die Vervollständigung der Lücken hilfreich sein:
<ul style="margin-bottom: 0px; margin-top: 0px">
    <li><code style="background-color: #FAEAEA; padding: 0px">scipy.signal.correlate2d</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.correlate2d.html" target="_blank">SciPy-Dokumentation</a><br>
        Diese Funktion ist für zweidimensionale Inputs (height, width) und für eine Verwendung ohne Bias konzipiert. Es gibt zudem auch eine Funktion scipy.signal.convolve2d,
welche passender klingt, die Filter allerdings vor der Anwendung um 180° dreht. Im nachfolgenden Code-Block wird die Faltungsoperation händisch realisiert.
    </li>
    <li><code style="background-color: #FAEAEA; padding: 0px">np.zeros</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://numpy.org/doc/stable/reference/generated/numpy.zeros.html" target="_blank">NumPy-Dokumentation</a>
    </li>
    <li><code style="background-color: #FAEAEA; padding: 0px">np.sum</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://numpy.org/doc/stable/reference/generated/numpy.sum.html" target="_blank">NumPy-Dokumentation</a>
    </li>
    <li><code style="background-color: #FAEAEA; padding: 0px">slice</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://docs.python.org/3/library/functions.html#slice" target="_blank">Python-Dokumentation</a>
        </li>
</ul>
</div>


*Implementierung:*

In [6]:
def conv2d(input_tensor, filter_tensor, bias_tensor=None, stride=(1, 1), padding=(0, 0), dilation=(1, 1), groups=1):
    # ueberpruefen, ob gueltige Parameter uebergeben wurden, andernfalls nicht implementiert als Fehler zurueckgeben
    assert stride == (1, 1), "Not implemented"
    assert padding == (0, 0), "Not implemented"
    assert dilation == (1, 1), "Not implemented"
    assert groups == 1, "Not implemented"

    # Groessen aus Groesse des Input Tensors extrahieren
    n_examples, n_channels_in, h_in, w_in = input_tensor.shape

    # Groessen aus Groesse des Filter Tensors extrahieren
    n_channels_out, n_channels_in_kernel, h_kernel, w_kernel = filter_tensor.shape

    # kleiner Check: sofern keine Grouped Convolution oder Depthwise Convolution sind beide identisch
    assert n_channels_in == n_channels_in_kernel, "Not implemented"

    # Groesse `(h_out, w_out)` der Output Feature Maps bestimmen
    h_out, w_out = get_output_shape(input_shape=(h_in,w_in), kernel_size=(h_kernel,w_kernel), 
                                    stride=stride, padding=padding, dilation=dilation) # bitte Code ergaenzen <---------------- [Luecke (3)]

    # leeren Output Tensor allokieren
    output_tensor = np.zeros((n_examples, n_channels_out, h_out, w_out), dtype=np.float32)

    # Output berechnen
    for ex in range(0, n_examples):
        # fuer jedes Beispiel im Batch

        for ch_out in range(0, n_channels_out):
            # fuer jeden Output Channel

            for i in range(0, h_out):
                # fuer jede y-Position i der Hoehe der Output Feature Maps

                for j in range(0, w_out):
                    # fuer jede x-Position j der Breite der Output Feature Maps

                    # Slices `vertical` und `horizontal` fuer relevanten Teil in den Input Feature Maps
                    # ausgehend von aktueller Position i und j bestimmen
                    vertical = slice(i,i+h_kernel) # bitte Code ergaenzen <---------------- [Luecke (4)]
                    horizontal = slice(j,j+w_kernel)# bitte Code ergaenzen <---------------- [Luecke (5)]

                    # Ergebnisse fuer jeden Input Channel akkumulieren
                    for ch_in in range(0, n_channels_in):
                        # für jeden Input Channel

                        # Teil der Input Feature Map `input_window` aus Input Tensor extrahieren
                        input_window = input_tensor[ex, ch_in, vertical, horizontal] # bitte Code ergaenzen <---------------- [Luecke (6)]

                        # 2D Filter `weight` aus Filter Tensor extrahieren
                        weight = filter_tensor[ch_out,ch_in,:,:] # bitte Code ergaenzen <---------------- [Luecke (7)]

                        # elementweise Multiplikation des Fenster mit Filtergewichten
                        # und anschliessende Summation
                        result = np.sum(input_window * weight, axis=None)

                        # Ergebnis zum bisherigen Ergebnis hinzuaddieren
                        output_tensor[ex, ch_out, i, j] += result

                    # nach der Faltung zur Feature Map gehoerendes Bias Gewicht addieren
                    if bias_tensor is not None:
                        output_tensor[ex,ch_out,i,j]+= bias_tensor[ch_out] # bitte Code ergaenzen <---------------- [Luecke (8)]

    return output_tensor

*Überprüfung:*

Zur Überprüfung können die Output Volumes aus der Aufgabenstellung berechnet werden.

In [7]:
# Netzwerkeingabe als Tensor mit Groesse 1x1x8x8 definieren
o_0 = np.array([[[[1.0, 1.0, 1.0, 0.0, 1.0, 1.0, 1.0, 1.0],
                  [1.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 1.0],
                  [1.0, 0.0, 1.0, 0.0, 1.0, 0.0, 1.0, 1.0],
                  [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 0.0],
                  [0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0],
                  [1.0, 1.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0],
                  [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
                  [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0]]]], dtype='float32')

# Gewichte Convolution 1
# Filter
w_1 = np.array([[[[2.0, -2.0],
                  [1.0, -1.0]]]], dtype='float32')
# Bias
b_1 = np.array([0], dtype='float32')

# Output Tensor nach der ersten Convolution mit Groesse 1x1x4x4 definieren
o_1 = np.array([[[[1.0, 2.0, 0.0, -1.0],
                  [3.0, 2.0, 1.0, 1.0],
                  [-2.0, 0.0, 1.0, 1.0],
                  [0.0, 1.0, 0.0, -1.0]]]], dtype='float32')

# Gewichte Convolution 2
# Filter
w_2 = np.array([[[[-2.0, -1.0],
                  [1.0, 2.0]]]], dtype='float32')
# Bias
b_2 = np.array([3], dtype='float32')


# Gewichte Convolution 3
# Filter
w_3 = np.array([[[[1.0, 1.0, 2.0],
                  [3.0, 0.0, 1.0],
                  [-1.0, 1.0, -2.0]]]], dtype='float32')
# Bias
b_3 = np.array([-2.5], dtype='float32')

# Forward Propagation: `o_2` und `o_3` berechnen
o_2 = conv2d(o_1,w_2,b_2) # bitte Code ergaenzen <---------------- [Luecke (9)]
o_3 = conv2d(o_2,w_3,b_3) # bitte Code ergaenzen <---------------- [Luecke (10)]

# Ergebnisse ausgeben
print_tensors(tensors=(o_2, o_3), labels=('o_2', 'o_3'))

o_2:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
o_3:
[[[[0.5]]]]


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
o_2:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
o_3:
[[[[0.5]]]]
</code>
</details>

---
<a name="gradienten"></a>
**In der Vorlesung und in der Übung wurde gezeigt, dass sich die Berechnungen der Gradienten für die Filter-Gewichte und das Input Volume ebenfalls auf Faltungsoperationen zurückführen lassen. Für die Gradienten der Bias-Gewichte ist hingegen keine Faltungsoperationen notwendig.**

Gradienten der Filter-Gewichte:
>Die Gradienten der Filter-Gewichte ergeben sich aus der Faltung des Input
Volumes mit den Gradienten des Output Volumes.

Gradienten des Input Volumes:
> Die Gradienten des Input Volumes ergeben sich aus der Faltung des mit
Zero Padding um (Filtergröße - 1) erweiterten Gradienten des Output Volumes mit
den um 180° gedrehten Filter-Gewichten.

Die Gradienten der Filtergewichte können folglich ohne Probleme mit der bereits implementierten Convolution berechnet werden. Für die Bestimmung der Gradienten des Input Volumes muss die Convolution so erweitert werden, dass der Parameter `padding` ebenfalls ausgewertet wird.
<br>
<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende Funktionen könnten für die Vervollständigung der Lücken hilfreich sein:
<ul style="margin-bottom: 0px; margin-top: 0px">
    <li><code style="background-color: #FAEAEA; padding: 0px">np.pad</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://numpy.org/doc/stable/reference/generated/numpy.pad.html" target="_blank">NumPy-Dokumentation</a>
        </li>
    <li><code style="background-color: #FAEAEA; padding: 0px">np.newaxis</code>&nbsp;&nbsp;&rarr;&nbsp;<a href="https://numpy.org/doc/stable/user/basics.indexing.html#dimensional-indexing-tools" target="_blank">NumPy-Dokumentation</a>
        </li>
</ul>
</div>

In [8]:
def conv2d_with_padding(input_tensor, filter_tensor, bias_tensor, stride=(1, 1), padding=(0, 0), dilation=(1, 1), groups=1):
    # Input Tensor mit Zero Padding zu `input_tensor_pad` erweitern
    input_tensor_pad = np.pad(input_tensor, 
                              pad_width=((0,0),(0,0),(padding[0],padding[0]),(padding[1],padding[1])),
                              mode = 'constant',
                              constant_values=0)# bitte Code ergaenzen <---------------- [Luecke (11)]

    # auf bereits implementierte Standard Convolution zurueckfuehren
    return conv2d(input_tensor_pad, filter_tensor, bias_tensor,
                  stride=stride,
                  padding=(0, 0),
                  dilation=dilation,
                  groups=groups)

Anschließend kann eine Klasse `StandardConvolution` zur Berechnung der Forward Propagation und Backpropagation implementiert werden. Die Realisierung in Form einer Klasse ermöglicht das Speichern der Gewichte sowie des Input-Tensors für die Backpropagation.

In [9]:
class StandardConvolution:
    def __init__(self, filter_tensor, bias_tensor):
        # Gewichte speichern
        self.weight = filter_tensor
        self.bias = bias_tensor

        # zum Speichern des Input Tensors
        self.input_tensor = None

    def forward(self, input_tensor):
        # Input Tensor für spaetere Backpropagation speichern
        self.input_tensor = input_tensor

        return conv2d(input_tensor, self.weight,
                      bias_tensor=self.bias,
                      stride=(1,1), 
                      padding=(0,0),
                      dilation=(1,1),
                      groups=1) # bitte Code ergaenzen <---------------- [Luecke (12)]

    def backward_bias(self, output_tensor_grad):
        # Gradienten ergeben sich durch Multiplikation mit 1 und anschließender Summation ueber
        # Batch-Achse sowie beiden raumlichen Achsen
        return (output_tensor_grad * 1).sum(axis=(0, 2, 3))

    def backward_weight(self, output_tensor_grad):
        # Gradienten ergeben sich aus der Faltung des Input Volumes mit den Gradienten des Output Volumes

        # leeren Tensor fuer Gradienten allokieren
        weight_grad = np.zeros_like(self.weight)

        # n_examples
        n_examples = self.input_tensor.shape[0]

        # die Gradienten ueber alle Beispiele im Batch aufaddieren
        for ex in range(0, n_examples):
            # fuer jedes Beispiel im Batch

            # Input Volume und Gradient des Output Volume extrahieren
            input_volume = self.input_tensor[ex]
            output_volume_grad = output_tensor_grad[ex]

            # Gradienten berechnen
            # Beachten Sie, dass der Gradient der Output Feature Map fuer alle Input Feature Maps
            # gleichermassen gilt. Durch eine Verschiebung der Channel-Achse auf die
            # Batch-Achse und einer kuenstlichen neuen Channel-Achse (np.newaxis), kann die
            # Berechnung fuer alle Input Feature Maps mit einer einzelnen Convolution
            # durchgefuehrt werden. Im Anschluss werden die beiden Achsen zurueck getauscht.
            conv2d_result = conv2d(input_volume[:, np.newaxis, :, :],
                                   output_volume_grad[:, np.newaxis, :, :],
                                   bias_tensor=None,
                                   stride=(1, 1),
                                   padding=(0, 0),
                                   dilation=(1, 1),
                                   groups=1)
            weight_grad += conv2d_result.swapaxes(0, 1)
        return weight_grad

    def backward_input(self, output_tensor_grad):
        # Gradienten ergeben sich aus der Faltung des mit Zero Padding um (Filtergröße - 1)
        # erweiterten Gradienten des Output Volumes mit den um 180 Grad gedrehten Filter-Gewichten

        # Groessen aus Groesse des Filter Tensors extrahieren
        n_channels_out, n_channels_in, h_kernel, w_kernel = self.weight.shape

        # Groessen aus Groesse des Gradienten Tensors extrahieren
        n_examples, _, h_output_grad, w_output_grad = output_tensor_grad.shape

        # Padding berechnen
        h_pad = h_kernel - 1
        w_pad = w_kernel - 1

        # Groesse fuer Gradienten berechnen
        h_input_grad, w_input_grad = \
            get_output_shape(input_shape=(h_output_grad, w_output_grad),
                                          kernel_size=(h_kernel, w_kernel),
                                          stride=(1, 1),
                                          padding=(h_pad, w_pad),
                                          dilation=(1, 1))

        # leeren Tensor fuer Gradienten allokieren
        input_grad = np.zeros((n_examples, n_channels_in, h_input_grad, w_input_grad),
                              dtype='float32')

        # Filter-Gewichte in den (height, width)-Dimensionen um 180 Grad drehen, bzw.
        # zweimal spiegeln (an height (2) und width (3) Achse):
        weight_rot180 = np.flip(np.flip(self.weight, axis=2), axis=3)

        # die Gradienten fuer jede Input Feature Map uber alle Output Feature Maps aufaddieren
        for ch_out in range(0, n_channels_out):
            # fuer jede Output Feature Map in allen Beispielen im Batch

            # Feature Maps und zugehoerige Gewichte extrahieren
            feature_maps_grad = output_tensor_grad[:, ch_out, :, :]
            weight_ch_out = weight_rot180[ch_out, :, :, :]

            # Gradienten berechnen
            # Beachten Sie, dass die Gradienten einer Output Feature Map fuer alle Beispiele
            # im Batch gleichzeitig berechnet werden. Hierzu werden kuenstliche neue
            # Channel-Achsen (np.newaxis) in den Tensoren ergaenzt.
            input_grad += conv2d_with_padding(feature_maps_grad[:, np.newaxis, :, :],
                                              weight_ch_out[:, np.newaxis, :, :],
                                              bias_tensor=None,
                                              stride=(1, 1),
                                              padding=(h_pad, w_pad),
                                              dilation=(1, 1),
                                              groups=1)

        return input_grad

*Überprüfung:*

Zur Überprüfung können nun die kompletten Ergebnisse für die letzten beiden Convolutions aus Teilaufgabe (a) reproduziert werden.

In [10]:
# Schichten anlegen
conv2 = StandardConvolution(w_2, b_2)
conv3 = StandardConvolution(w_3, b_3)

# Forward Propagation: `o_2` und `o_3` berechnen
o_2 = conv2.forward(o_1) # bitte Code ergaenzen <---------------- [Luecke (13)]
o_3 = conv3.forward(o_2) # bitte Code ergaenzen <---------------- [Luecke (14)]
y = o_3.squeeze()    # zusaetzliche Achsen entfernen

# Fehler bestimmen
t = 0
e_mse = (y-t)**2
print_tensors(tensors=(o_2, o_3, e_mse), labels=('o_2', 'o_3', 'E'))

o_2:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
o_3:
[[[[0.5]]]]
E:
0.25


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
o_2:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
o_3:
[[[[0.5]]]]
E:
0.25
</code>
</details>

In [11]:
eta = 0.0005

# Backpropagation Fehler
dedy = 2*(y-t)
dedo_3 = dedy.reshape(o_3.shape)    # zusaetzliche Achsen wiederherstellen

# Backpropagation Convolution 3: `dedb_3`, `dedw_3` und `dedo_2` bestimmen
dedb_3 = conv3.backward_bias(dedo_3) # bitte Code ergaenzen <---------------- [Luecke (15)]
dedw_3 = conv3.backward_weight(dedo_3) # bitte Code ergaenzen <---------------- [Luecke (16)]
dedo_2 = conv3.backward_input(dedo_3) # bitte Code ergaenzen <---------------- [Luecke (17)]

b_3_new = b_3 - eta*dedb_3
w_3_new = w_3 - eta*dedw_3

print_tensors(tensors=(dedb_3, dedw_3, dedo_2, b_3_new, w_3_new),
              labels=('dEdb_3', 'dEdw_3', 'dEdo_2', 'b_3_new', 'w_3_new'),
              precision=4)

# Backpropagation Convolution 2: `dedb_2`, `dedw_2` und `dedo_1` bestimmen
dedb_2 = conv2.backward_bias(dedo_2) # bitte Code ergaenzen <---------------- [Luecke (18)]
dedw_2 = conv2.backward_weight(dedo_2) # bitte Code ergaenzen <---------------- [Luecke (19)]
dedo_1 = conv2.backward_input(dedo_2) # bitte Code ergaenzen <---------------- [Luecke (20)]

b_2_new = b_2 - eta*dedb_2
w_2_new = w_2 - eta*dedw_2

print_tensors(tensors=(dedb_2, dedw_2, dedo_1, b_2_new, w_2_new),
              labels=('dEdb_2', 'dEdw_2', 'dEdo_1', 'b_2_new', 'w_2_new'),
              precision=4)

interactive_check('backpropagation')

dEdb_3:
[1.]
dEdw_3:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
dEdo_2:
[[[[ 1.  1.  2.]
   [ 3.  0.  1.]
   [-1.  1. -2.]]]]
b_3_new:
[-2.5005]
w_3_new:
[[[[ 0.997   0.9985  1.9965]
   [ 3.0035  0.      0.9985]
   [-1.0045  0.9985 -1.999 ]]]]
dEdb_2:
[6.]
dEdw_2:
[[[[13.  6.]
   [ 3.  7.]]]]
dEdo_1:
[[[[-2. -3. -5. -2.]
   [-5.  0.  2.  3.]
   [ 5.  5.  4.  4.]
   [-1. -1.  0. -4.]]]]
b_2_new:
[2.997]
w_2_new:
[[[[-2.0065 -1.003 ]
   [ 0.9985  1.9965]]]]


Gradient 'dedb_3':


Gradient 'dedw_3':


Gradient 'dedo_2':


Neue Bias-Gewichte 'b_3_new':


Neue Filter-Gewichte 'w_3_new':


Gradient 'dedb_2':


Gradient 'dedw_2':


Gradient 'dedo_1':


Neue Bias-Gewichte 'b_2_new':


Neue Filter-Gewichte 'w_2_new':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
dEdb_3:
[1.]
dEdw_3:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
dEdo_2:
[[[[ 1.  1.  2.]
   [ 3.  0.  1.]
   [-1.  1. -2.]]]]
b_3_new:
[-2.5005]
w_3_new:
[[[[ 0.997   0.9985  1.9965]
   [ 3.0035  0.      0.9985]
   [-1.0045  0.9985 -1.999 ]]]]
dEdb_2:
[6.]
dEdw_2:
[[[[13.  6.]
   [ 3.  7.]]]]
dEdo_1:
[[[[-2. -3. -5. -2.]
   [-5.  0.  2.  3.]
   [ 5.  5.  4.  4.]
   [-1. -1.  0. -4.]]]]
b_2_new:
[2.997]
w_2_new:
[[[[-2.0065 -1.003 ]
   [ 0.9985  1.9965]]]]
</code>
</details>

<hr style="border-width: 5px">

<a name="praktikum"></a>
<h3 style="color: #aa0000;">Praktikumsaufgabe P6.1: Strided Convolution</h3>

Beziehen Sie zusätzlich noch die Convolution der ersten Schicht (Strided Convolution) mit in die Betrachtungen ein und reproduzieren Sie die Ergebnisse aus Teilaufgabe (a) ebenfalls für diese Convolution.

---
**Erweitern Sie hierzu zunächst die Funktion `conv2d` so, dass ebenfalls Strided Convolutions realisiert werden können.**

*Implementierung:*

In [12]:
def conv2d(input_tensor, filter_tensor, bias_tensor=None, stride=(1, 1), padding=(0, 0), dilation=(1, 1), groups=1):
    # ueberpruefen, ob gueltige Parameter uebergeben wurden, andernfalls nicht implementiert als Fehler zurueckgeben
    assert padding == (0, 0), "Not implemented"
    assert dilation == (1, 1), "Not implemented"
    assert groups == 1, "Not implemented"

    # Groessen aus Groesse des Input Tensors extrahieren
    n_examples, n_channels_in, h_in, w_in = input_tensor.shape

    # Groessen aus Groesse des Filter Tensors extrahieren
    n_channels_out, n_channels_in_kernel, h_kernel, w_kernel = filter_tensor.shape

    # kleiner Check: sofern keine Grouped Convolution oder Depthwise Convolution sind beide identisch
    assert n_channels_in == n_channels_in_kernel, "Not implemented"

    # Groesse `(h_out, w_out)` der Output Feature Maps bestimmen
    h_out, w_out = get_output_shape(input_shape=(h_in, w_in),
                                    kernel_size=(h_kernel, w_kernel),
                                    stride=stride,
                                    padding=padding,
                                    dilation=dilation)

    # leeren Output Tensor allokieren
    output_tensor = np.zeros((n_examples, n_channels_out, h_out, w_out), dtype=np.float32)

    # Output berechnen
    for ex in range(0, n_examples):
        # fuer jedes Beispiel im Batch

        for ch_out in range(0, n_channels_out):
            # fuer jeden Output Channel

            for i in range(0, h_out):
                # fuer jede y-Position i der Hoehe der Output Feature Maps

                for j in range(0, w_out):
                    # fuer jede x-Position j der Breite der Output Feature Maps

                    # Slices `vertical` und `horizontal` fuer relevanten Teil in den Input Feature Maps
                    # ausgehend von aktueller Position i und j bestimmen
                    vertical = slice(i,i+h_kernel) # bitte Code ergaenzen <---------------- [Luecke (21)]
                    horizontal = slice(j,j+w_kernel) # bitte Code ergaenzen <---------------- [Luecke (22)]

                    # Ergebnisse fuer jeden Input Channel akkumulieren
                    for ch_in in range(0, n_channels_in):
                        # für jeden Input Channel

                        # Teil der Input Feature Map `input_window` aus Input Tensor extrahieren
                        input_window = input_tensor[ex, ch_in, vertical, horizontal]

                        # 2D Filter `weight` aus Filter Tensor extrahieren
                        weight = filter_tensor[ch_out, ch_in, :, :]

                        # elementweise Multiplikation des Fenster mit Filtergewichten
                        # und anschliessende Summation
                        result = np.sum(input_window * weight, axis=None)

                        # Ergebnis zum bisherigen Ergebnis hinzuaddieren
                        output_tensor[ex, ch_out, i, j] += result

                    # nach der Faltung zur Feature Map gehoerendes Bias Gewicht addieren
                    if bias_tensor is not None:
                        output_tensor[ex, ch_out, i, j] += bias_tensor[ch_out]

    return output_tensor

*Überprüfung:*

In [29]:
# Forward Propagation: `o_1` berechnen
o_1 = conv2d(o_0,w_1,b_1,stride=(2,2)) # bitte Code ergaenzen <---------------- [Luecke (23)]

# Ergebnis ausgeben
print_tensors(tensors=o_1, labels='o_1')

# Ergebnis ueberpruefen
interactive_check('conv1_forward')

o_1:
[[[[ 1. -1.  2. -2.]
   [ 3. -3.  1. -1.]
   [ 3. -2.  2. -2.]
   [ 1.  1.  0.  0.]]]]


Output Tensor 'o_1':


<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
o_1:
[[[[ 1.  2.  0. -1.]
   [ 3.  2.  1.  1.]
   [-2.  0.  1.  1.]
   [ 0.  1.  0. -1.]]]]
</code>
</details>

---
**Implementieren Sie anschließend eine Klasse `StridedConvolution`, mit der alle fehlenden Ergebnisse und die Gradienten in Richtung der Netzwerkeingabe `o_0` bestimmt werden können.**
<br>
<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Neben den Betrachtungen in der Übungsaufgabe, könnte folgende Literaturstelle für die Vervollständigung der Lücken hilfreich sein:
<ul style="margin-bottom: 0px; margin-top: 0px">
    <li>Dumoulin, et. al.: <i>A guide to convolution arithmetic for deep learning</i>, arXiv 2016.&nbsp;&nbsp;&rarr;&nbsp;<a href="https://arxiv.org/pdf/1603.07285.pdf" target="_blank">arXiv</a>
        </li>
</ul>
</div>

*Implementierung:*

Die Realisierung wird erfordern, dass an verschiedenen Stellen eine Dilation eingebracht werden muss.
Hierzu können Sie die nachfolgende Hilfsfunktion verwenden:

In [14]:
def add_dilation(tensor, dilation=(1, 1)):
    d0, d1, h, w = tensor.shape
    h_r, w_r = dilation

    tensor_dilated = np.zeros((d0, d1, (h-1)*h_r + 1, (w-1)*w_r +1),
                              dtype='float32')
    tensor_dilated[:, :, ::h_r, ::w_r] = tensor
    return tensor_dilated

Zur Veranschaulichung der Funktionsweise soll folgendes Beispiel dienen:

In [15]:
tensor = np.arange(1, 10).reshape((1, 1, 3, 3))
tensor_r2 = add_dilation(tensor, dilation=(2, 2))

print_tensors(tensors=(tensor, tensor_r2),
              labels=('Tensor', 'Tensor mit Dilation r=2'))

Tensor:
[[[[1 2 3]
   [4 5 6]
   [7 8 9]]]]
Tensor mit Dilation r=2:
[[[[1. 0. 2. 0. 3.]
   [0. 0. 0. 0. 0.]
   [4. 0. 5. 0. 6.]
   [0. 0. 0. 0. 0.]
   [7. 0. 8. 0. 9.]]]]


Realisierung der Klasse `StridedConvolution`:

In [16]:
class StridedConvolution(StandardConvolution):
    def __init__(self, filter_tensor, bias_tensor, stride):
        super(StridedConvolution, self).__init__(filter_tensor, bias_tensor)

        # Stride speichern
        self.stride = stride

    def forward(self, input_tensor):
        # Input Tensor für spaetere Backpropagation speichern
        self.input_tensor = input_tensor

        # Anpassungen fuer StridedConvolution
        return # bitte Code ergaenzen <---------------- [Luecke (24)]

    def backward_weight(self, output_tensor_grad):
        # Anpassungen fuer StridedConvolution
        # bitte Code ergaenzen <---------------- [Luecke (25)]

        # Implementierung in Basisklasse wiederverwenden
        return super(StridedConvolution, self).backward_weight(output_tensor_grad)

    def backward_input(self, output_tensor_grad):
        # Anpassungen fuer StridedConvolution
        # bitte Code ergaenzen <---------------- [Luecke (26)]

        # Implementierung in Basisklasse wiederverwenden
        return super(StridedConvolution, self).backward_input(output_tensor_grad)

*Überprüfung:*

In [17]:
# Schicht anlegen
conv1 = StridedConvolution(w_1, b_1, stride=(2, 2))

# Forward Propagation Convolution 1
o_1 = conv1.forward(o_0)

# Backpropagation Convolution 1: `dedb_1`, `dedw_1` und `dedo_0` bestimmen
# bitte Code ergaenzen <---------------- [Luecke (27)]
# bitte Code ergaenzen <---------------- [Luecke (28)]
# bitte Code ergaenzen <---------------- [Luecke (29)]

b_1_new = b_1 - eta*dedb_1
w_1_new = w_1 - eta*dedw_1

# Ergebnisse ausgeben
print_tensors(tensors=(dedb_1, dedw_1, dedo_0, b_1_new, w_1_new),
              labels=('dEdb_1', 'dEdw_1', 'dEdo_0', 'b_1_new', 'w_1_new'),
              precision=4)

# Ergebnisse ueberpruefen
interactive_check('conv1_backward')

NameError: name 'dedb_1' is not defined

<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
dEdb_1:
[0.]
dEdw_1:
[[[[-14.  -3.]
   [ -1.  -8.]]]]
dEdo_0:
[[[[ -4.   4.  -6.   6. -10.  10.  -4.   4.]
   [ -2.   2.  -3.   3.  -5.   5.  -2.   2.]
   [-10.  10.   0.   0.   4.  -4.   6.  -6.]
   [ -5.   5.   0.   0.   2.  -2.   3.  -3.]
   [ 10. -10.  10. -10.   8.  -8.   8.  -8.]
   [  5.  -5.   5.  -5.   4.  -4.   4.  -4.]
   [ -2.   2.  -2.   2.   0.   0.  -8.   8.]
   [ -1.   1.  -1.   1.   0.   0.  -4.   4.]]]]
b_1_new:
[0.]
w_1_new:
[[[[ 2.007  -1.9985]
   [ 1.0005 -0.996 ]]]]
</code>
</details>

<hr style="border-width: 5px">

### (g) Vergleichen Sie die Ergebnisse Ihrer NumPy-Implementierung mit einer PyTorch-Implementierung.

---
Pakete importieren:

In [ ]:
# PyTorch
import torch
import torch.nn.functional as F

*Netzwerk implementieren:*

In [ ]:
class CNN(torch.nn.Module):
    def __init__(self, print_tensors=True):
        super(CNN, self).__init__()

        # Schicht 1 `self.conv1` anlegen
        # bitte Code ergaenzen <---------------- [Luecke (30)]

        # Schicht 2 `self.conv2` anlegen
        # bitte Code ergaenzen <---------------- [Luecke (31)]

        # Schicht 3 `self.conv3` anlegen
        # bitte Code ergaenzen <---------------- [Luecke (32)]

        # Gewichte mit bereits definierten Variablen initialisieren
        self.conv1.weight.data = torch.tensor(w_1)
        self.conv1.bias.data = torch.tensor(b_1)
        self.conv2.weight.data = torch.tensor(w_2)
        self.conv2.bias.data = torch.tensor(b_2)
        self.conv3.weight.data = torch.tensor(w_3)
        self.conv3.bias.data = torch.tensor(b_3)

        self.print_tensors = print_tensors

    def forward(self, x):
        # Convolution 1: `o_1` bestimmen
        # bitte Code ergaenzen <---------------- [Luecke (33)]

        # Convolution 2: `o_2` bestimmen
        # bitte Code ergaenzen <---------------- [Luecke (34)]

        # Convolution 3: `o_3` bestimmen
        # bitte Code ergaenzen <---------------- [Luecke (35)]

        y = o_3.view(-1, 1)

        if self.print_tensors:
            x.register_hook(lambda grad: print_tensors(grad, 'o_0.grad'))
            o_1.register_hook(lambda grad: print_tensors(grad, 'o_1.grad'))
            o_2.register_hook(lambda grad: print_tensors(grad, 'o_2.grad'))
            o_3.register_hook(lambda grad: print_tensors(grad, 'o_3.grad'))
            print_tensors(tensors=(o_1, o_2, o_3),
                          labels=('o_1', 'o_2', 'o_3'))

        return y

---

*Netzwerkobjekt anlegen, auf Eingabe anwenden und Fehler berechnen:*

In [ ]:
# Netzwerkobjekt anlegen
network = CNN()

# Eingabe konvertieren
o_0_pytorch = torch.tensor(o_0, requires_grad=True)

# Netzwerkobjekt auf Eingabe anwenden
y = network(o_0_pytorch)

# Fehler berechnen und ausgeben
t = torch.tensor([[0]], dtype=torch.float32)
e = F.mse_loss(y, t)

print_tensors(tensors=e, labels='E')

# Backpropagation
e.backward()

# Gradienten ausgeben
for n, p in network.named_parameters():
    print_tensors(tensors=p.grad, labels=f"{n}.grad")

<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
o_1:
[[[[ 1.  2.  0. -1.]
   [ 3.  2.  1.  1.]
   [-2.  0.  1.  1.]
   [ 0.  1.  0. -1.]]]]
o_2:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
o_3:
[[[[0.5]]]]
E:
0.25
o_3.grad:
[[[[1.]]]]
o_2.grad:
[[[[ 1.  1.  2.]
   [ 3.  0.  1.]
   [-1.  1. -2.]]]]
o_1.grad:
[[[[-2. -3. -5. -2.]
   [-5.  0.  2.  3.]
   [ 5.  5.  4.  4.]
   [-1. -1.  0. -4.]]]]
o_0.grad:
[[[[ -4.   4.  -6.   6. -10.  10.  -4.   4.]
   [ -2.   2.  -3.   3.  -5.   5.  -2.   2.]
   [-10.  10.   0.   0.   4.  -4.   6.  -6.]
   [ -5.   5.   0.   0.   2.  -2.   3.  -3.]
   [ 10. -10.  10. -10.   8.  -8.   8.  -8.]
   [  5.  -5.   5.  -5.   4.  -4.   4.  -4.]
   [ -2.   2.  -2.   2.   0.   0.  -8.   8.]
   [ -1.   1.  -1.   1.   0.   0.  -4.   4.]]]]
conv1.weight.grad:
[[[[-14.  -3.]
   [ -1.  -8.]]]]
conv1.bias.grad:
[0.]
conv2.weight.grad:
[[[[13.  6.]
   [ 3.  7.]]]]
conv2.bias.grad:
[6.]
conv3.weight.grad:
[[[[ 6.  3.  7.]
   [-7.  0.  3.]
   [ 9.  3. -2.]]]]
conv3.bias.grad:
[1.]
</code>
</details>

<hr style="border-width: 5px">

### (h) Führen Sie in Ihrer Implementierung eine weitere Forward Propagation mit den aktualisierten Gewichten durch. Wird mit den neuen Gewichten ein geringerer Fehler erzielt?

In [ ]:
# Netzwerkobjekt anlegen
network = CNN(print_tensors=False)

# Optimierer anlegen
optimizer = torch.optim.SGD(network.parameters(), lr=0.0005, momentum=0.0, nesterov=False)

# Eingabe konvertieren
o_0_pytorch = torch.tensor(o_0, requires_grad=False)

for epoch in range(2):
    # Netzwerkobjekt auf Eingabe anwenden
    y = network(o_0_pytorch)

    # Fehler berechnen und ausgeben
    t = torch.tensor([[0]], dtype=torch.float32)
    e = F.mse_loss(y, t)
    print_tensors(tensors=e, labels=f'E (t={epoch+1})', precision=4)

    # Backpropagation: Gradienten bestimmen und Lernschritt ausfuehren
    # bitte Code ergaenzen <---------------- [Luecke (36)]
    # bitte Code ergaenzen <---------------- [Luecke (37)]

<details>
    <summary>&#9432; <i>Überprüfung &nbsp; &nbsp; <font color="CCCCCC">(anklicken, um Lösung anzuzeigen)</font></i></summary>
<code style="padding: 0px">
E (t=1):
0.25
E (t=2):
0.0083
</code>
</details>

<hr style="border-width: 5px">

<a name="praktikum2"></a>
<h3 style="color: #aa0000;">Praktikumsaufgabe P6.2: Convolution mit Dilation & Padding</h3>

Leiten Sie anhand systematischer Experimente ab, welche Ableitungen sich ausgehend von den Fehlergradienten am Output $\dfrac{\partial E}{\partial O^{(l)}}$ einer Schicht in Richtung der Filter-Gewichte $W^{(l)}$ und des Inputs $O^{(l-1)}$ bei Verwendung der nachfolgenden Convolutions ergeben.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende Literaturstelle könnte für die Vervollständigung der Lücken hilfreich sein:
<ul style="margin-bottom: 0px; margin-top: 0px">
    <li>Dumoulin, et. al.: <i>A guide to convolution arithmetic for deep learning</i>, arXiv 2016.&nbsp;&nbsp;&rarr;&nbsp;<a href="https://arxiv.org/pdf/1603.07285.pdf" target="_blank">arXiv</a>
        </li>
</ul>
</div>

Beachten Sie, dass die für die Faltungen realisierten Klassen stets von der nachfolgenden Klasse (`ConvolutionBase`) ableiten sollten - diese Basisklasse untersützt bereits eine Funktion zur Realisierung einer Faltungsoperation (`self.conv2d`) mit allen bekannten Parametern. Es ist daher nicht notwendig, die oben im Notebook verwendeten Funktionen händisch zu erweitern.

In [ ]:
from tui_dl4cv.cnn import ConvolutionBase

Folgende Funktionen werden an verschiedenen Stellen benötigt und untersützen Sie bei der Realisierung:

In [ ]:
def rotate_180(tensor):
    """Tensor entlang der raeumlichen Dimensionen um 180 Grad drehen"""
    return np.flip(np.flip(tensor, axis=2), axis=3)

def swap_axes(tensor, axis1=0, axis2=1):
    """Zwei Achsen innerhalb eines Tensors tauschen"""
    return np.swapaxes(tensor, axis1=axis1, axis2=axis2)

Nutzen Sie zudem die vorgesehene Überprüfung zur Validierung Ihrer Impelemtierung.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
<b>Hinweis</b>: Falls einige der Tests fehlschlegen, wird eine entsprechende Fehlermeldung ausgegeben. Beachten Sie, dass die Fehler in '[Optional]' und '[Obligatorisch]' unterteilt sind. Damit Sie die volle Punktzahl erhalten, dürfen keine mit '[Obligatorisch]' gekennzeichneten Fehler auftreten.</div>

---
**Dilated Convolution:**

\begin{equation}
\text{Kanalanzahl Input } n^{(l-1)} = 1\\[4mm]
\text{Kanalanzahl Output } n^{(l)} = 1, \; \text{Filtergröße } k^{(l)} = 2, \; \text{Stride } s^{(l)} = 1, \; \text{Padding } p^{(l)} = 0, \; \text{Dilation } r^{(l)} = 2, \; \text{Groups } g^{(l)} = 1
\end{equation}

Beschreiben Sie <u>ganz kurz</u>, wie die Convolution parametriert und durchgeführt werden muss.
Untermauern Sie Ihre Argumentationen durch das Einsetzen geeigneter Parameter in nachfolgenden Code.

<span style="color: #ff0000;">(Anpassung der Antwort durch Doppelklick hier)</span>

In [ ]:
class DilatedConvolution(ConvolutionBase):
    def __init__(self):
        # oben definierte Parameter weitergeben, im Nachgang ebenfalls als self.* verfuegbar
        super().__init__(
            in_channels=1,
            out_channels=1,
            kernel_size=2,
            stride=1,
            padding=0,
            dilation=2,
            groups=1,
            bias=False
        )

    def backward_input(self, output_tensor_grad):
        weight_tensor = self.weight

        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (38)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return self.conv2d(input_, filter_,
                           stride=stride,
                           padding=padding,
                           dilation=dilation,
                           groups=groups)

    def backward_weight(self, output_tensor_grad):
        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (39)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return swap_axes(self.conv2d(input_, filter_,
                                     stride=stride,
                                     padding=padding,
                                     dilation=dilation,
                                     groups=groups))

# Objekt der Schicht erstellen
conv = DilatedConvolution()

# exemparische Anwendung (fuer Testzwecke und Verstaendnisexperimente)
x = np.ones((1, 1, 4, 4))
y = conv(x)
dedy = np.ones_like(y)
dedw = conv.backward_weight(dedy)
dedx = conv.backward_input(dedy)
print_tensors(tensors=(conv.weight, x, y, dedy, dedw, dedx),
              labels=('weight', 'x', 'y', 'dedy', 'dedw', 'dedx'))

# Impelementierung ueberpruefen
interactive_check('dilated_conv')

---
**Convolution mit Padding:**

\begin{equation}
\text{Kanalanzahl Input } n^{(l-1)} = 1\\[4mm]
\text{Kanalanzahl Output } n^{(l)} = 1, \; \text{Filtergröße } k^{(l)} = 3, \; \text{Stride } s^{(l)} = 1, \; \text{Padding } p^{(l)} = 1, \; \text{Dilation } r^{(l)} = 1, \; \text{Groups } g^{(l)} = 1
\end{equation}

Beschreiben Sie <u>ganz kurz</u>, wie die Convolution parametriert und durchgeführt werden muss.
Untermauern Sie Ihre Argumentationen durch das Einsetzen geeigneter Parameter in nachfolgenden Code.

<span style="color: #ff0000;">(Anpassung der Antwort durch Doppelklick hier)</span>

In [ ]:
class ConvolutionWithPadding(ConvolutionBase):
    def __init__(self):
        # oben definierte Parameter weitergeben, im Nachgang ebenfalls als self.* verfuegbar
        super().__init__(
            in_channels=1,
            out_channels=1,
            kernel_size=3,
            stride=1,
            padding=1,
            dilation=1,
            groups=1,
            bias=False
        )

    def backward_input(self, output_tensor_grad):
        weight_tensor = self.weight

        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (40)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return self.conv2d(input_, filter_,
                           stride=stride,
                           padding=padding,
                           dilation=dilation,
                           groups=groups)

    def backward_weight(self, output_tensor_grad):
        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (41)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return swap_axes(self.conv2d(input_, filter_,
                                     stride=stride,
                                     padding=padding,
                                     dilation=dilation,
                                     groups=groups))

# Objekt der Schicht erstellen
conv = ConvolutionWithPadding()

# exemparische Anwendung (fuer Testzwecke und Verstaendnisexperimente)
# siehe DilatedConvolution

# Impelementierung ueberpruefen
interactive_check('conv_with_padding')

<hr style="border-width: 5px">

<a name="praktikum3"></a>
<h3 style="color: #aa0000;">Praktikumsaufgabe P6.3: Convolution mit 1x1-Filtern & Gruppen</h3>

Leiten Sie anhand systematischer Experimente ab, welche Ableitungen sich ausgehend von den Fehlergradienten am Output $\dfrac{\partial E}{\partial O^{(l)}}$ einer Schicht in Richtung der Filter-Gewichte $W^{(l)}$ und des Inputs $O^{(l-1)}$ bei Verwendung der nachfolgenden Convolutions ergeben.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Folgende Literaturstelle könnte für die Vervollständigung der Lücken hilfreich sein:
<ul style="margin-bottom: 0px; margin-top: 0px">
    <li>Dumoulin, et. al.: <i>A guide to convolution arithmetic for deep learning</i>, arXiv 2016.&nbsp;&nbsp;&rarr;&nbsp;<a href="https://arxiv.org/pdf/1603.07285.pdf" target="_blank">arXiv</a>
        </li>
</ul>
</div>

Beachten Sie, dass die für die Faltungen realisierten Klassen stets von der nachfolgenden Klasse (`ConvolutionBase`) ableiten sollten - diese Basisklasse untersützt bereits eine Funktion zur Realisierung einer Faltungsoperation (`self.conv2d`) mit allen bekannten Parametern. Es ist daher nicht notwendig, die oben im Notebook verwendeten Funktionen händisch zu erweitern.

In [ ]:
from tui_dl4cv.cnn import ConvolutionBase

Folgende Funktionen werden an verschiedenen Stellen benötigt und untersützen Sie bei der Realisierung:

In [ ]:
def rotate_180(tensor):
    """Tensor entlang der raeumlichen Dimensionen um 180 Grad drehen"""
    return np.flip(np.flip(tensor, axis=2), axis=3)

def swap_axes(tensor, axis1=0, axis2=1):
    """Zwei Achsen innerhalb eines Tensors tauschen"""
    return np.swapaxes(tensor, axis1=axis1, axis2=axis2)

Nutzen Sie zudem die vorgesehene Überprüfung zur Validierung Ihrer Impelemtierung.

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
<b>Hinweis</b>: Falls einige der Tests fehlschlegen, wird eine entsprechende Fehlermeldung ausgegeben. Beachten Sie, dass die Fehler in '[Optional]' und '[Obligatorisch]' unterteilt sind. Damit Sie die volle Punktzahl erhalten, dürfen keine mit '[Obligatorisch]' gekennzeichneten Fehler auftreten.</div>

---
**1x1 Convolution mit zunehmender Anzahl an Kanälen:**

\begin{equation}
\text{Kanalanzahl Input } n^{(l-1)} = 1\\[4mm]
\text{Kanalanzahl Output } n^{(l)} = 2, \; \text{Filtergröße } k^{(l)} = 1, \; \text{Stride } s^{(l)} = 1, \; \text{Padding } p^{(l)} = 0, \; \text{Dilation } r^{(l)} = 1, \; \text{Groups } g^{(l)} = 1
\end{equation}

Beschreiben Sie <u>ganz kurz</u>, wie die Convolution parametriert und durchgeführt werden muss.
Untermauern Sie Ihre Argumentationen durch das Einsetzen geeigneter Parameter in nachfolgenden Code.

<span style="color: #ff0000;">(Anpassung der Antwort durch Doppelklick hier)</span>

In [ ]:
class Convolution1x1MoreChannels(ConvolutionBase):
    def __init__(self):
        # oben definierte Parameter weitergeben, im Nachgang ebenfalls als self.* verfuegbar
        super().__init__(
            in_channels=1,
            out_channels=2,
            kernel_size=1,
            stride=1,
            padding=0,
            dilation=1,
            groups=1,
            bias=False
        )

    def backward_input(self, output_tensor_grad):
        weight_tensor = self.weight

        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (42)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return self.conv2d(input_, filter_,
                           stride=stride,
                           padding=padding,
                           dilation=dilation,
                           groups=groups)

    def backward_weight(self, output_tensor_grad):
        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (43)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return swap_axes(self.conv2d(input_, filter_,
                                     stride=stride,
                                     padding=padding,
                                     dilation=dilation,
                                     groups=groups))

# Objekt der Schicht erstellen
conv = Convolution1x1MoreChannels()

# exemparische Anwendung (fuer Testzwecke und Verstaendnisexperimente)
# siehe DilatedConvolution

# Impelementierung ueberpruefen
interactive_check('conv_1x1_more_channels')

---
**1x1 Convolution mit Reduktion der Kanalanzahl:**

\begin{equation}
\text{Kanalanzahl Input } n^{(l-1)} = 2\\[4mm]
\text{Kanalanzahl Output } n^{(l)} = 1, \; \text{Filtergröße } k^{(l)} = 1, \; \text{Stride } s^{(l)} = 1, \; \text{Padding } p^{(l)} = 0, \; \text{Dilation } r^{(l)} = 1, \; \text{Groups } g^{(l)} = 1
\end{equation}

Beschreiben Sie <u>ganz kurz</u>, wie die Convolution parametriert und durchgeführt werden muss.
Untermauern Sie Ihre Argumentationen durch das Einsetzen geeigneter Parameter in nachfolgenden Code.

<span style="color: #ff0000;">(Anpassung der Antwort durch Doppelklick hier)</span>

In [ ]:
class Convolution1x1LessChannels(ConvolutionBase):
    def __init__(self):
        # oben definierte Parameter weitergeben, im Nachgang ebenfalls als self.* verfuegbar
        super().__init__(
            in_channels=2,
            out_channels=1,
            kernel_size=1,
            stride=1,
            padding=0,
            dilation=1,
            groups=1,
            bias=False
        )

    def backward_input(self, output_tensor_grad):
        weight_tensor = self.weight

        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (44)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return self.conv2d(input_, filter_,
                           stride=stride,
                           padding=padding,
                           dilation=dilation,
                           groups=groups)

    def backward_weight(self, output_tensor_grad):
        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (45)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return swap_axes(self.conv2d(input_, filter_,
                                     stride=stride,
                                     padding=padding,
                                     dilation=dilation,
                                     groups=groups))

# Objekt der Schicht erstellen
conv = Convolution1x1LessChannels()

# exemparische Anwendung (fuer Testzwecke und Verstaendnisexperimente)
# siehe DilatedConvolution

# Impelementierung ueberpruefen
interactive_check('conv_1x1_less_channels')

---
**Grouped Convolution:**

\begin{equation}
\text{Kanalanzahl Input } n^{(l-1)} = 2\\[4mm]
\text{Kanalanzahl Output } n^{(l)} = 4, \; \text{Filtergröße } k^{(l)} = 3, \; \text{Stride } s^{(l)} = 1, \; \text{Padding } p^{(l)} = 1, \; \text{Dilation } r^{(l)} = 1, \; \text{Groups } g^{(l)} = 2
\end{equation}

<br>
<div style="background-color: #FAEAEA; padding: 5px; margin: 5px 0px 5px 0px; border-radius: 5px;">
Beachten Sie, dass die Implementierung für die Grouped Convolution anspruchsvoller als die anderen Implementierungen ist. Für beide Gradientenberechnungen ist im Backward Pass ebenfalls eine Grouped Convolution nötig. Jedoch müssen die Gewichte in <code>backward_input</code> bzw. der Input in <code>backward_weight</code> durch <code>np.reshape</code> und <code>swap_axes</code> zuvor entsprechend umstrukturiert werden.</div>

Beschreiben Sie <u>ganz kurz</u>, wie die Convolution parametriert und durchgeführt werden muss.
Untermauern Sie Ihre Argumentationen durch das Einsetzen geeigneter Parameter in nachfolgenden Code.

<span style="color: #ff0000;">(Anpassung der Antwort durch Doppelklick hier)</span>

In [ ]:
class GroupedConvolution(ConvolutionBase):
    def __init__(self):
        # oben definierte Parameter weitergeben, im Nachgang ebenfalls als self.* verfuegbar
        super().__init__(
            in_channels=2,
            out_channels=4,
            kernel_size=3,
            stride=1,
            padding=1,
            dilation=1,
            groups=2,
            bias=False
        )

    def backward_input(self, output_tensor_grad):
        weight_tensor = self.weight

        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (46)]
        stride =
        padding =
        dilation =
        groups =

        input_ =
        filter_ =





        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        return self.conv2d(input_, filter_,
                           stride=stride,
                           padding=padding,
                           dilation=dilation,
                           groups=groups)

    def backward_weight(self, output_tensor_grad):
        # Parameter und Eingaben fuer Faltung definieren
        # bitte Code ergaenzen <---------------- [Luecke (47)]
        stride =
        padding =
        dilation =
        groups =

        input_ =




        filter_ =

        # Faltungsoperation durchfuehren (sollte nicht veraendert werden)
        result = swap_axes(self.conv2d(input_, filter_,
                                       padding=padding,
                                       stride=stride,
                                       dilation=dilation,
                                       groups=groups))
        return result

# Objekt der Schicht erstellen
conv = GroupedConvolution()

# exemparische Anwendung (fuer Testzwecke und Verstaendnisexperimente)
# siehe DilatedConvolution

# Impelementierung ueberpruefen
interactive_check('grouped_conv')

$_{_\text{Created for Deep Learning for Computer Vision (DL4CV)}}$